File: fetch/fetch_test.go
package fetch

import (
	"fmt"
	"net/http"
	"net/http/httptest"
	"strings"
	"testing"

	"github.com/binzume/goja_utils"
	"github.com/dop251/goja"
)

func newTestServer() *httptest.Server {
	var sampleHandler = http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		if r.URL.Path != "/" {
			http.Error(w, "Not Found.", http.StatusNotFound)
			return
		}

		fmt.Fprintf(w, `{"status": "ok"}`)
	})
	return httptest.NewServer(sampleHandler)
}

func TestFetch(t *testing.T) {
	ts := newTestServer()
	defer ts.Close()

	r := goja_utils.NewJsRunnner()

	// Enables Fetch API
	r.Run(func(vm *goja.Runtime) {
		Enable(vm)
		vm.GlobalObject().Set("testServerUrl", ts.URL)
	})

	r.Start()

	ret, err := r.RunFile("fetch_test.js")
	if err != nil {
		t.Fatalf("Failed to load %v", err)
	}

	resultValue, err2 := r.Await(ret)
	if err2 != nil {
		t.Fatalf("Failed to test: %v", err2)
	}

	result := resultValue.Export()
	if str, ok := result.(string); ok && strings.TrimSpace(str) == "pass" {
		t.Log(str)
	} else {
		t.Errorf("Unexpected result: %v", result)
	}
}

File: fetch/fetch_test.js
const baseUrl = globalThis.testServerUrl || "http://binzume.net";
console.log("Test url:", baseUrl);

async function test_get_text() {
    let res = await fetch(baseUrl);
    if (res.status != 200) {
        throw "res.status != 200";
    }
    let text = await res.text();
    if (!text) {
        throw "!text";
    }
}

async function test_get_bytes() {
    let res = await fetch(baseUrl);
    if (res.status != 200) {
        throw "res.status != 200";
    }
    let data = await res.bytes();
    if (!data instanceof Uint8Array) {
        throw `!data instanceof Uint8Array`;
    }
}

async function test_get_json() {
    let res = await fetch(baseUrl);
    if (res.status != 200) {
        throw "res.status != 200";
    }
    let json = await res.json();
    if (json?.status != "ok") {
        throw "json?.status != ok";
    }
}

async function test_404() {
    let res = await fetch(baseUrl + "/_test_not_found");
    if (res.status != 404) {
        throw "res.status != 404";
    }
    let text = await res.text();
    if (!text) {
        throw "!text";
    }
}

async function test() {
    await test_get_text();
    await test_get_bytes();
    await test_get_json();
    await test_404();

    return "pass"
}

test();
